Fix make targets

chrisgilmerproj opened the pull request:
## Description

This supports work I'm doing in #1923  

It's really hard to build individual targets with our old paradigm. I've broken out all the individual go files into separate targets. I've also either remove unnecessary stamp files or targets that slowed things down. Where possible I tried to use Make for what it's good at: recognizing if files are built or not.

## Setup

Here are a couple good targets to try:

```sh
make clean build
make e2e_test_docker
```

Honestly you ought to try as many make targets as possible.

## Code Review Verification Steps

* [x] Deploy to Experimental
* [x] Request review from a member of a different team.

chrisgilmerproj commented on Makefile:
```diff
@@ -245,7 +303,7 @@ server_run_default: server_deps server_generate db_dev_run
 	$(AWS_VAULT) ./bin/gin \
 		--build ./cmd/milmove \
-		--bin /bin/milmove \
+		--bin /bin/milmove_gin \
```
I noticed a race condition where if you're running the server AND the docker e2e tests at the same time you may try to overwrite `bin/milmove`. Renaming this here is a noop and ensures we don't have a race condition.

chrisgilmerproj commented on Makefile:
```diff
@@ -258,30 +316,33 @@ server_run_debug:
+	bin/make-tsp-user \
+	bin/save-fuel-price-data \
+	bin/send-to-gex \
+	bin/tsp-award-queue
```
The main difference here is that if you call `make build_tools build_tools` it won't rebuild everything if those targets exist. You'd have to do `make build_tools && rm -rf bin/ && make build_tools` to get it to build everything again.

tinyels replied:
This is good intermediate step, but I would love it if the tools had targets that did proper dependency checking. Or at least the most commonly used targets. For instance, `generate-test-data`.

chrisgilmerproj replied:
I hope that I've captured dependency checking in the targets above this. If I missed one let me know. But you should be able to do `make bin/generate-test-data` now and it ought to check all the dependencies first.

chrisgilmerproj commented on Makefile:
```diff
@@ -258,30 +316,33 @@ server_run_debug:
-	go build -i -ldflags "$(LDFLAGS)" -o bin/tsp-award-queue ./cmd/tsp_award_queue
+build_tools: .bash_version.stamp \
+	server_deps \
+	server_generate \
```
I think I ought to move `server_deps server_generate` to all of the individual targets instead of having it here. Because we want to be able to rebuild specific targets too and those may rely on swagger.

jim commented:
I'm in favor of this change. It makes the `Makefile` easier to understand.

rdhariwal commented on Makefile:
```diff
@@ -70,24 +70,37 @@ else
+	check_bash_version \
+	ensure_pre_commit \
+	client_deps \
+	server_deps
```
love the readability

chrisgilmerproj commented:
> I was purposefully trying to find a gap in the dependencies and this throws an error:
> 
> ```
> $ make clean
> $ make bin/make-tsp-user
> scripts/gen-server
> scripts/gen-server: line 9: ./bin/swagger: No such file or directory
> make: *** [.server_generate.stamp] Error 127
> ```

Awesome find! I'm also running through this right now.  I'll correct it and look for others.

chrisgilmerproj commented:
> I got an error with the following after `make clean`.
> 
> ```
> mymove (cg_fix_make_targets) $ make spellcheck
> node_modules/.bin/mdspell --ignore-numbers --ignore-acronyms --en-us \
> 		`find . -type f -name "*.md" \
> 			-not -path "./node_modules/*" \
> 			-not -path "./docs/adr/index.md"`
> /bin/sh: node_modules/.bin/mdspell: No such file or directory
> ```
> 
> Should we add a make target for `node_modules`, install mdspell directly, or add a warning note saying what to install before? `make client_build` seems to install the necessary dependency, but might want to break it down further.

We have `.client_deps.stamp` as a target, which is good enough. I've updated some of these random targets to fix their dependencies and removed the ones that no one uses.

chrisgilmerproj commented:
Deployed to experimental and it was a success! https://circleci.com/workflow-run/cc6d51be-1a69-465e-821a-abc5b78dfab8
